File: Frontend/src/api/patientService.js
// src/api/patientService.js

import api from "./axios";

// Get all patients
export const getAllPatients = () =>
  api.get("/patients");

// Get patient by ID
export const getPatientById = (id) =>
  api.get(`/patients/${id}`);

// Register a new patient
export const createPatient = (data) =>
  api.post("/patients", data);

// Update patient details
export const updatePatient = (id, data) =>
  api.put(`/patients/${id}`, data);

// Delete patient
export const deletePatient = (id) =>
  api.delete(`/patients/${id}`);

// Search patients by keyword (name, phone, etc.)
export const searchPatients = (keyword) =>
  api.get("/patients/search", { params: { keyword } });

File: Backend/com.demo/src/main/java/com/demo/service/PatientService.java
package com.demo.service;

import com.demo.model.Patient;
import com.demo.repository.PatientRepository;
import com.demo.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
@RequiredArgsConstructor
public class PatientService {

    private final PatientRepository patientRepository;

    public List<Patient> getAllPatients() {
        return patientRepository.findByActive(true);
    }

    public Patient getPatientById(Long id) {
        return patientRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Patient not found with id: " + id));
    }

    public Patient createPatient(Patient patient) {
        patient.setActive(true); // ensure new patients are active
        return patientRepository.save(patient);
    }

    public Patient updatePatient(Long id, Patient updated) {
        Patient patient = getPatientById(id);
        patient.setFirstName(updated.getFirstName());
        patient.setLastName(updated.getLastName());
        patient.setDateOfBirth(updated.getDateOfBirth());
        patient.setGender(updated.getGender());
        patient.setBloodGroup(updated.getBloodGroup());
        patient.setPhone(updated.getPhone());
        patient.setEmail(updated.getEmail());
        patient.setAddress(updated.getAddress());
        patient.setMedicalHistory(updated.getMedicalHistory());
        patient.setAllergies(updated.getAllergies());
        return patientRepository.save(patient);
    }

    public void deletePatient(Long id) {
        Patient patient = getPatientById(id);
        patient.setActive(false);
        patientRepository.save(patient); 
    }
    
 // Search patients by first name or last name
    public List<Patient> searchPatients(String keyword) {
        return patientRepository
            .findByFirstNameContainingOrLastNameContaining(keyword, keyword);
    }
}